mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -0,0 +1,622 @@
+
+        self.client.connect(&opts).await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn reconnect_and_subscribe(&self) -> MmResult<(), WalletConnectError> {
+        self.connect_client().await?;
+        // Resubscribes to previously active session topics after reconnection.
+        let sessions = self
+            .session_manager
+            .get_sessions()
+            .flat_map(|s| vec![s.topic, s.pairing_topic])
+            .collect::<Vec<_>>();
+
+        if !sessions.is_empty() {
+            self.client.batch_subscribe(sessions).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Create a WalletConnect pairing connection url.
+    pub async fn new_connection(
+        &self,
+        required_namespaces: serde_json::Value,
+        optional_namespaces: Option<serde_json::Value>,
+    ) -> MmResult<String, WalletConnectError> {
+        let required_namespaces = serde_json::from_value(required_namespaces)?;
+        let optional_namespaces = match optional_namespaces {
+            Some(value) => Some(serde_json::from_value(value)?),
+            None => None,
+        };
+        let (topic, url) = self.pairing.create(self.metadata.clone(), None)?;
+
+        info!("[{topic}] Subscribing to topic");
+
+        for attempt in 0..MAX_RETRIES {
+            match self
+                .client
+                .subscribe(topic.clone())
+                .timeout_secs(PUBLISH_TIMEOUT_SECS)
+                .await
+            {
+                Ok(res) => {
+                    res.map_to_mm(|err| err.into())?;
+                    info!("[{topic}] Subscribed to topic");
+                    send_proposal_request(self, &topic, required_namespaces, optional_namespaces).await?;
+                    return Ok(url);
+                },
+                Err(_) => self.wait_until_client_is_online_loop(attempt).await,
+            }
+        }
+
+        MmError::err(WalletConnectError::InternalError(
+            "client connection timeout".to_string(),
+        ))
+    }
+
+    /// Retrieves the symmetric key associated with a given `topic`.
+    fn sym_key(&self, topic: &Topic) -> MmResult<SymKey, WalletConnectError> {
+        self.session_manager
+            .sym_key(topic)
+            .or_else(|| self.pairing.sym_key(topic).ok())
+            .ok_or_else(|| {
+                error!("Failed to find sym_key for topic: {topic}");
+                MmError::new(WalletConnectError::InternalError(format!(
+                    "topic sym_key not found: {topic}"
+                )))
+            })
+    }
+
+    /// Handles an inbound published message by decrypting, decoding, and processing it.
+    async fn handle_published_message(&self, msg: PublishedMessage) -> MmResult<(), WalletConnectError> {
+        let message = {
+            let key = self.sym_key(&msg.topic)?;
+            decode_and_decrypt_type0(msg.message.as_bytes(), &key)?
+        };
+
+        info!("[{}] Inbound message payload={message}", msg.topic);
```

Comment:
since we print the message i guess this fits more as a debug log, so to not flood the logs